src/main/java/picard/analysis/CollectUmiPrevalenceMetrics.java

```diff
@@ -0,0 +1,181 @@
+    public Integer MINIMUM_BARCODE_BQ = 30;
+
+    @Argument(shortName = "FUR", doc = "Whether to filter unpaired reads from the input.", optional = true)
+    public boolean FILTER_UNPAIRED_READS = true;
+
+    @Argument(
+            fullName = "PROGRESS_STEP_INTERVAL",
+            doc = "The interval between which progress will be displayed.",
+            optional = true
+    )
+    public int PROGRESS_STEP_INTERVAL = 1_000_000;
+
+
+    private static final Log log = Log.getInstance(CollectUmiPrevalenceMetrics.class);
+
+    @Override
+    protected int doWork() {
+
+        IOUtil.assertFileIsReadable(INPUT);
+
+        // get an iterator to reads that overlap the heterozygous sites
+        final Histogram<Integer> umiCount = new Histogram<>("numUmis", "duplicateSets");
+        final CountingPairedFilter countingPairedFilter = new CountingPairedFilter();
+        final CountingFilterWrapper countingAlignedFilter = new CountingFilterWrapper(new AlignedFilter(true));
+        final CountingMapQFilter countingMapQFilter = new CountingMapQFilter(MINIMUM_MQ);
+        final CountingFilterWrapper countingSecondaryOrSupplementaryFilter =
+                new CountingFilterWrapper(new SecondaryOrSupplementaryFilter());
+        final ProgressLogger progress = new ProgressLogger(log, PROGRESS_STEP_INTERVAL, "examined", "duplicate sets");
+
+        try (SamReader in = SamReaderFactory.makeDefault()
+                .referenceSequence(REFERENCE_SEQUENCE)
+                .open(INPUT)) {
+
+            IOUtil.assertFileIsWritable(OUTPUT);
+
+            final SAMRecordIterator samRecordIterator = in.iterator();
+            final List<SamRecordFilter> samFilters = CollectionUtil.makeList(
+                    countingAlignedFilter,
+                    countingMapQFilter,
+                    countingSecondaryOrSupplementaryFilter
+            );
+            if (FILTER_UNPAIRED_READS) {
+                samFilters.add(countingPairedFilter);
+            }
+
+            final FilteringSamIterator filteredSamRecordIterator = new FilteringSamIterator(samRecordIterator, new AggregateFilter(samFilters));
+            log.info("Queried BAM, getting duplicate sets.");
+
+            // get duplicate iterator from iterator above
+            final DuplicateSetIterator duplicateSets = new DuplicateSetIterator(filteredSamRecordIterator, in.getFileHeader(), false, null, log);
+
+            log.info("Starting iteration on duplicate sets");
+
+            set:
+            while (duplicateSets.hasNext()) {
+
+                final DuplicateSet set = duplicateSets.next();
+                final SAMRecord setRep = set.getRepresentative();
+
+                progress.record(setRep);
+                Set<String> barcodes = new HashSet<>();
+                for (final SAMRecord read : set.getRecords()) {
+                    if (!read.hasAttribute(BARCODE_TAG)) {
+                        log.warn("no barcode tag!");
```

Comment:
Would it make sense to bake these checks into filters instead of doing them here?

Reply:
sure.